apollo-federation/src/merger/merger.rs

```diff
@@ -828,8 +830,235 @@ impl Merger {
+    ) -> Result<HashMap<FieldDefinitionPosition, Sources<()>>, FederationError> {
+        todo!("Implement add_fields_shallow")
+    }
+
+    fn validate_override(
+        &self,
+        sources: &Sources<()>,
+        dest: &FieldDefinitionPosition,
+    ) -> Result<FieldMergeContext, FederationError> {
+        todo!("Implement validate_override")
+    }
+
+    fn validate_subscription_field(
+        &mut self,
+        sources: &Sources<()>,
+        dest: &FieldDefinitionPosition,
+    ) -> Result<(), FederationError> {
+        // no subgraph marks field as @shareable
+        let mut fields_with_shareable: Sources<Node<FieldDefinition>> = Default::default();
+        for (idx, unit) in sources.iter() {
+            if unit.is_some() {
+                let subgraph = &self.subgraphs[*idx];
+                let shareable_directive_name = &subgraph
+                    .metadata()
+                    .federation_spec_definition()
+                    .shareable_directive_definition(subgraph.schema())?
+                    .name;
+                if dest.has_applied_directive(subgraph.schema(), shareable_directive_name) {
+                    let field = dest.get(subgraph.schema().schema())?;
+                    fields_with_shareable.insert(*idx, Some(field.node.clone()));
+                }
+            }
+        }
+        if !fields_with_shareable.is_empty() {
+            self.error_reporter
+                .add_error(CompositionError::InvalidFieldSharing {
+                    message:
+                        "Fields on root level subscription object cannot be marked as shareable"
+                            .to_string(),
+                    locations: self.source_locations(&fields_with_shareable),
+                });
+        }
+        Ok(())
+    }
+
+    fn are_all_fields_external(
+        &self,
+        idx: usize,
+        ty: &ObjectOrInterfaceTypeDefinitionPosition,
+    ) -> Result<bool, FederationError> {
+        Ok(ty.fields(self.merged.schema())?.all(|field| {
+            self.subgraphs[idx]
+                .metadata()
+                .external_metadata()
+                .is_external(&FieldDefinitionPosition::from(field.clone()))
+        }))
+    }
+
+    fn hint_on_inconsistent_value_type_field(
+        &mut self,
+        sources: &Sources<()>,
+        dest: &ObjectOrInterfaceTypeDefinitionPosition,
+        field: &FieldDefinitionPosition,
+    ) -> Result<(), FederationError> {
+        let (hint_id, type_description) = match field {
+            FieldDefinitionPosition::Object(field) => (
+                HintCode::InconsistentObjectValueTypeField,
+                "non-entity object",
+            ),
+            FieldDefinitionPosition::Interface(field) => (
+                HintCode::InconsistentInterfaceValueTypeField,
+                "non-entity interface",
+            ),
+            _ => panic!("Expected field to be an Object or Interface field, but it is not"),
```

Comment:
We can return an internal error instead of crashing here
```suggestion
            _ => bail!("Expected field to be an Object or Interface field, but it is not"),
```